components/script/dom/gpu.rs

```diff
@@ -0,0 +1,168 @@
+
+use crate::compartments::InCompartment;
+use crate::dom::bindings::codegen::Bindings::GPUBinding::GPURequestAdapterOptions;
+use crate::dom::bindings::codegen::Bindings::GPUBinding::{self, GPUMethods, GPUPowerPreference};
+use crate::dom::bindings::error::Error;
+use crate::dom::bindings::refcounted::{Trusted, TrustedPromise};
+use crate::dom::bindings::reflector::{reflect_dom_object, DomObject, Reflector};
+use crate::dom::bindings::root::DomRoot;
+use crate::dom::bindings::str::DOMString;
+use crate::dom::globalscope::GlobalScope;
+use crate::dom::gpuadapter::GPUAdapter;
+use crate::dom::promise::Promise;
+use crate::task::TaskOnce;
+use dom_struct::dom_struct;
+use ipc_channel::ipc::{self, IpcSender};
+use ipc_channel::router::ROUTER;
+use js::jsapi::Heap;
+use std::rc::Rc;
+use std::sync::{Arc, Mutex};
+use webgpu::wgpu;
+use webgpu::{WebGPURequest, WebGPUResponse, WebGPUResponseResult, WebGPUThread};
+
+#[dom_struct]
+pub struct GPU {
+    reflector_: Reflector,
+}
+
+impl GPU {
+    pub fn new_inherited() -> GPU {
+        GPU {
+            reflector_: Reflector::new(),
+        }
+    }
+
+    pub fn new(global: &GlobalScope) -> DomRoot<GPU> {
+        reflect_dom_object(Box::new(GPU::new_inherited()), global, GPUBinding::Wrap)
+    }
+
+    fn get_wgpu_thread(&self) -> Option<WebGPUThread> {
+        self.global().as_window().webgpu_thread()
+    }
+}
+
+pub trait AsyncWGPUListener {
+    fn handle_response(&self, response: WebGPUResponse, promise: &Rc<Promise>);
+}
+
+struct WGPUResponse<T: AsyncWGPUListener + DomObject> {
+    promise: Option<TrustedPromise>,
+    receiver: Trusted<T>,
+}
+
+impl<T: AsyncWGPUListener + DomObject> WGPUResponse<T> {
+    #[allow(unrooted_must_root)]
+    fn response(&mut self, response: WebGPUResponseResult) {
+        let promise = self.promise.take().expect("promise is missing").root();
+
+        // JSAutoRealm needs to be manually made.
+        // Otherwise, Servo will crash.
+        match response {
+            Ok(response) => self.receiver.root().handle_response(response, &promise),
+            Err(error) => promise.reject_error(Error::Type(format!(
+                "Received error from WebGPU thread: {}",
+                error
+            ))),
+        }
+    }
+}
+
+pub fn response_async<T: AsyncWGPUListener + DomObject + 'static>(
+    promise: &Rc<Promise>,
+    receiver: &T,
+) -> IpcSender<WebGPUResponseResult> {
+    let (action_sender, action_receiver) = ipc::channel().unwrap();
+    let task_source = receiver.global().networking_task_source();
+    let context = Arc::new(Mutex::new(WGPUResponse {
+        promise: Some(TrustedPromise::new(promise.clone())),
+        receiver: Trusted::new(receiver),
+    }));
+    ROUTER.add_route(
+        action_receiver.to_opaque(),
+        Box::new(move |message| {
+            struct ListenerTask<T: AsyncWGPUListener + DomObject> {
+                context: Arc<Mutex<WGPUResponse<T>>>,
```

Comment:
I don't think you need the mutex and the arc, here, because `WGPUResponse` itself will be `Send`, and I don't think you need to make it `Sync` to use it as you intend in `TaskOnce`.

Reply:
Looks like the `Sync` is needed here:
```
error[E0277]: `*const dom::promise::Promise` cannot be shared between threads safely
  --> components/script/dom/gpu.rs:91:21
   |
91 |             impl<T> TaskOnce for ListenerTask<T>
   |                     ^^^^^^^^ `*const dom::promise::Promise` cannot be shared between threads safely
   |
   = help: within `dom::gpu::WGPUResponse<T>`, the trait `std::marker::Sync` is not implemented for `*const dom::promise::Promise`
   = note: required because it appears within the type `dom::bindings::refcounted::TrustedPromise`
   = note: required because it appears within the type `std::option::Option<dom::bindings::refcounted::TrustedPromise>`
   = note: required because it appears within the type `dom::gpu::WGPUResponse<T>`
   = note: required because of the requirements on the impl of `std::marker::Send` for `std::sync::Arc<dom::gpu::WGPUResponse<T>>`
   = note: required because it appears within the type `dom::gpu::response_async::{{closure}}#0::ListenerTask<T>
```

Reply:
I think you are getting this error because of the `Arc<dom::gpu::WGPUResponse<T>>`, which I think is unnecessary, you can try to use the `WGPUResponse<T>` directly and I think it should work.

Reply:
I think I've figured it out in: https://github.com/servo/servo/pull/24708/commits/7f4ccba4dd7f65884006d439528913e344fb12f8